internal/providers/msentraid/msentraid.go

```diff
@@ -368,19 +418,80 @@ func (p Provider) VerifyUsername(requestedUsername, authenticatedUsername string
 	return nil
 }
 
+// SupportsDeviceRegistration checks if the provider supports device registration.
+func (p *Provider) SupportsDeviceRegistration() bool {
+	// The Microsoft Entra ID provider supports device registration.
+	return true
+}
+
+// IsTokenForDeviceRegistration checks if the token is for device registration.
+func (p *Provider) IsTokenForDeviceRegistration(token *oauth2.Token) (bool, error) {
+	accessToken, _, err := new(jwt.Parser).ParseUnverified(token.AccessToken, jwt.MapClaims{})
+	if err != nil {
+		return false, fmt.Errorf("failed to parse access token: %v", err)
+	}
+
+	appID, err := p.getAppID(accessToken)
+	if err != nil {
+		return false, fmt.Errorf("failed to get app ID from access token: %v", err)
+	}
+
+	return appID == consts.MicrosoftBrokerAppID, nil
+}
+
+// MaybeRegisterDevice checks if the device is already registered and registers it if not.
+func (p *Provider) MaybeRegisterDevice(ctx context.Context, token *oauth2.Token, username, issuerURL string, jsonData []byte) ([]byte, error) {
+	// If this function is called, it means that the token that we have is for device registration,
+	// so we can't use it to access the Microsoft Graph API.
+	p.needsAccessTokenForGraphAPI = true
+
+	// Check if the device is already registered
+	if len(jsonData) > 0 {
+		var data deviceRegistrationData
+		if err := json.Unmarshal(jsonData, &data); err != nil {
+			log.Noticef(ctx, "Device registration JSON data: %s", string(jsonData))
+			return nil, fmt.Errorf("failed to unmarshal device registration data: %v", err)
+		}
+		if data.DeviceID != "" && data.CertKey != nil && data.TransportKey != nil && data.AuthValue != "" && data.TPMMachineKey != nil {
+			return jsonData, nil
+		}
+	}
+
+	// Extract the tenant ID from the issuer URL
+	// https://login.microsoftonline.com/8de88d99-6d0f-44d7-a8a5-925b012e5940/v2.0
+	// -> 8de88d99-6d0f-44d7-a8a5-925b012e5940
+	tenantID := strings.Split(strings.TrimPrefix(issuerURL, "https://login.microsoftonline.com/"), "/")[0]
+	nameParts := strings.Split(username, "@")
+	if len(nameParts) != 2 {
+		return nil, fmt.Errorf("invalid username format: %s, expected format is 'username@domain'", username)
+	}
+	domain := nameParts[1]
```

Comment:
since it's the standard library, maybe `mail.ParseAddress` too?

Reply:
the `Address` struct returned by `mail.ParseAddress` doesn't even expose the domain field, so what would we gain by using it?

Reply:
It's to ensure the value *is* an email, I also was wondering if it could provide us more, but at least it's a formal check on the format we expect.